### PR TITLE
(Ragged OD Pipeline Step 1/3) Support RaggedImages and Bounding Boxes in BaseImageAugmentationLayer

### PR DESCRIPTION
**Pull request by LukeWood**
# What does this PR do?

This layer adds support for ragged images & cleans up ragged bounding box support.

layers.RandomFlip() will now support ragged images.

One API delta is that we always output ragged bounding boxes, but this is our native format so it is reasonable to do so.

---

another tiny delta that got included - adds a batchwise argument to MaybeApply so we can use it with CutMix & MixUp

### CODE REVIEW

**Comment by bhack** [comment]
Can we have only random flip here?

**Comment by LukeWood** [comment]
> Can we have only random flip here?

not really, the updates to base layer are not compatible with old layer versions and its easier to batch update.

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py [hunk 1]
```diff
@@ -107,6 +105,24 @@ def augment_image(self, image, transformation):
     def __init__(self, seed=None, **kwargs):
         super().__init__(seed=seed, **kwargs)
 
+    @property
+    def force_output_ragged_images(self):
+        """Control whether to force outputting of ragged images."""
+        return getattr(self, "_force_output_ragged_images", False)
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/base_image_augmentation_layer.py:111]: What's the tradeoff between using these settable properties vs a constructor argument?

I feel like `force_ragged=True` or similar could be a nice construction-time control of this

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py [hunk 2]
```diff
@@ -129,12 +145,101 @@ def __init__(self):
...
+            ragged_spec = tf.RaggedTensorSpec(
+                shape=images.shape[1:],
+                ragged_rank=1,
+                dtype=self.compute_dtype,
+            )
+            return ragged_spec
+        return tf.TensorSpec(images.shape[1:], self.compute_dtype)
+
+    # TODO(lukewood): promote to user facing API if needed
+    def _compute_bounding_box_signature(self, bounding_boxes):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: Do I understand correctly that this is forcing bounding boxes to be ragged? If so -- is that too coercive of an API?

File: keras_cv/layers/preprocessing/random_flip.py [hunk 3]
```diff
@@ -152,6 +155,9 @@ def augment_bounding_boxes(
                 "`RandomFlip(bounding_box_format='xyxy')`"
             )
 
+        if isinstance(bounding_boxes, tf.RaggedTensor):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/random_flip.py:158]: Why does this need to be made dense here?

File: keras_cv/layers/preprocessing/resizing.py [hunk 4]
```diff
@@ -59,6 +59,8 @@ class Resizing(BaseImageAugmentationLayer):
             largest possible resize of the image (of size `(height, width)`) that
             matches the target aspect ratio. By default
             (`pad_to_aspect_ratio=False`), aspect ratio may not be preserved.
+        pad_only: If True, resize the images without aspect ratio distortion only using
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/resizing.py:62]: Maybe this API change should be a separate PR?

File: keras_cv/layers/preprocessing/maybe_apply.py [hunk 5]
```diff
@@ -29,6 +29,8 @@ class MaybeApply(BaseImageAugmentationLayer):
         rate: controls the frequency of applying the layer. 1.0 means all elements in
             a batch will be modified. 0.0 means no elements will be modified.
             Defaults to 0.5.
+        batchwise: (Optional) bool, whether or not to pass entire batches to the
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/maybe_apply.py:32]: Is the expected behavior here that if batchwise=true, we decide on a per-batch level whether to apply a layer?

If so, might be worth documenting that.

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:111]: So the reason you don't want to do this as a constructor argument is that there is NO user use case to tweak this argument.

A given layer ALWAYS has this set, or not set.  That's why we made `auto_vectorize` an attribute as well.  Basically the goal is to fully hide this from users.

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: Yeah, so right now this is always making bounding boxes Ragged.

I'm on the fence - on one hand we can minimize API drift by requiring bounding boxes to be RaggedTensors.  on the other hand it is a bit more coercive.  

We can also internally always process raggeds, and then handle outputs being converted to Dense tensors in `format_output` and preserve the input format.  The issue with this is that if we just keep converting back and forth we're not netting ANY performance gains; its a performance hit.  

I say we submit this as is, and if users request dense bounding box tensors we implement it then.   Thoughts?

File: keras_cv/layers/preprocessing/maybe_apply.py (hunk 5 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/maybe_apply.py:32]: Updated the docstring, updated the behavior, etc

File: keras_cv/layers/preprocessing/random_flip.py (hunk 3 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/random_flip.py:158]: Getting the shape of a RaggedTensor is inconsistent, and it doesn't work if we don't make the tensor dense here.  The tensor is always dense at this point.

In a previous iteration this was required - but I just tested without it and it works now.  The base layer handles this now.

File: keras_cv/layers/preprocessing/resizing.py (hunk 4 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/resizing.py]: Removed - yeah good point its probably not the right move

**Comment by LukeWood** [comment]
Ok @ianstenbit all comments addressed or justification provided

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 2 above)
Comment (ianstenbit) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: I'm really not sure what the right answer here is. Maybe @martin-gorner has an opinion?

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py [hunk 6]
```diff
@@ -339,16 +466,27 @@ def _format_inputs(self, inputs):
...
 
+        valid_keys = [
+            IMAGES,
+            LABELS,
+            TARGETS,
+            BOUNDING_BOXES,
+            KEYPOINTS,
+            SEGMENTATION_MASKS,
+        ]
+        if not all([x in valid_keys for x in inputs.keys()]):
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/base_image_augmentation_layer.py:477]: I think this is a departure for the current functionality where we just pass-through all the keys we don't know about. What's the value of raising an error here?

Is the idea to alert users of a typo if e.g. they had a key like "buonding_boxes"?

I worry a bit that this is too coercive, but at the same time I can imagine that using a KPL and not having the boxes augmented because of a typo in your dict keys would be very annoying.

File: keras_cv/layers/preprocessing/ragged_image_test.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        {
+            "height": 224,
+            "width": 224,
+            "zoom_factor": (0.8, 1.0),
+            "aspect_ratio_factor": (3 / 4, 4 / 3),
+        },
+    ),
+]
+
+RAGGED_OUTPUT_TEST_CONFIGURATIONS = []
```
Comment (ianstenbit) [inline keras_cv/layers/preprocessing/ragged_image_test.py:109]: Is there a missing test case for these configurations / should this be removed?

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 6 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py]: Reverted

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 1 above)
Comment (tanzhenyu) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:111]: +1 -- how does user know which layers should force ragged output, and which should be false?

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 2 above)
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: any reason why this is private method, but `compute_image_signature` is public method?

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py [hunk 8]
```diff
@@ -129,12 +145,101 @@ def __init__(self):
...
+
+        return fn_output_signature
+
+    @staticmethod
+    def _any_ragged(inputs):
+        if isinstance(inputs[IMAGES], tf.RaggedTensor):
+            return True
+        if BOUNDING_BOXES in inputs:
+            return True
+        # TODO(lukewood): exhaustively check cases
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/base_image_augmentation_layer.py:227]: do we have to check for keypoints as well?

File: keras_cv/layers/preprocessing/grayscale.py [hunk 9]
```diff
@@ -53,6 +53,18 @@ def __init__(self, output_channels=1, **kwargs):
         # This layer may raise an error when running on GPU using auto_vectorize
         self.auto_vectorize = False
 
+    def compute_image_signature(self, images):
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/grayscale.py:56]: maybe comment on why overriding `compute_image_signature` is required?

File: keras_cv/layers/preprocessing/mosaic.py [hunk 10]
```diff
@@ -234,10 +238,12 @@ def _update_bounding_box(
             images=images,
             dtype=self.compute_dtype,
         )
-
         boxes_for_mosaic = tf.gather(bounding_boxes, permutation_order[index])
+        if isinstance(boxes_for_mosaic, tf.RaggedTensor):
+            boxes_for_mosaic = boxes_for_mosaic.to_tensor(-1, shape=[None, 5])
+        # boxes_for_mosaic = boxes_for_mosaic.to_tensor(-1)
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/mosaic.py:244]: remove if not needed?

File: keras_cv/layers/preprocessing/mosaic_test.py [hunk 11]
```diff
@@ -80,7 +80,7 @@ def test_return_shapes(self):
 
         self.assertEqual(xs.shape, [2, 512, 512, 3])
         self.assertEqual(ys_labels.shape, [2, 10])
-        self.assertEqual(ys_bounding_boxes.shape, [2, 12, 5])
+        self.assertEqual(ys_bounding_boxes.shape, [2, None, 5])
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/mosaic_test.py:83]: any reason this is changed to None?

**Review by tanzhenyu — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: Right now I see no use case to customize the bounding box output signature

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 8 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:227]: Done

File: keras_cv/layers/preprocessing/mosaic_test.py (hunk 11 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/mosaic_test.py:83]: It outputs a ragged instead of a padded dense now after the API change - this is an expected change.

Comment (LukeWood) [reply keras_cv/layers/preprocessing/mosaic_test.py:83]: (we could call to_tensor() in the test if we want)

**Comment by LukeWood** [comment]
Final round of comments addressed, @ianstenbit and @tanzhenyu PTAL and LGTM if it is good after the previous batch of comments.

File: keras_cv/layers/preprocessing/base_image_augmentation_layer.py (hunk 2 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/base_image_augmentation_layer.py:167]: Lets add an API for this if users ask for it.  We'll be processing Raggeds internally no matter what.

**Review by ianstenbit — APPROVED** [review]
Thanks Luke!

File: keras_cv/layers/preprocessing/resizing.py [hunk 12]
```diff
@@ -182,6 +194,7 @@ def resize_single_with_pad_to_aspect(x):
             inputs["images"] = image
 
             if boxes is not None:
+                boxes = keras_cv.bounding_box.filter_sentinels(boxes)
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/resizing.py:197]: unrelated to this PR: filter_empty_boxes?

File: keras_cv/layers/preprocessing/randomly_zoomed_crop.py [hunk 13]
```diff
@@ -79,7 +79,7 @@ def __init__(
         )
 
         self._check_class_arguments(height, width, zoom_factor, aspect_ratio_factor)
-
+        self.force_output_dense_images = True
```
Comment (tanzhenyu) [inline keras_cv/layers/preprocessing/randomly_zoomed_crop.py:82]: this seems to be introducing more user cognitive load, is there any other way to reduce it?
